scripts/enforce-backplane-rules.py

```diff
@@ -0,0 +1,316 @@
+                    failed = True
+
+                    self.log_failure(
+                         path,
+                         f"SubjectPermission '{name}' has clusterPermission(s) '{', '.join(invalid_perms)}'"
+                         " without the '-cluster' suffix"
+                    )
+
+               if invalid_perms := self._invalid_permission_names(subject_permission):
+                    failed = True
+
+                    self.log_failure(
+                         path,
+                         f"SubjectPermission '{name}' has permission(s) '{', '.join(invalid_perms)}'"
+                         " without the '-project' suffix"
+                    )
+
+          return failed
+
+     def _invalid_cluster_permission_names(self, subject_permission):
+          cluster_permissions = subject_permission.content['spec'].get('clusterPermissions', [])
+          unknown_role_names = self._filter_known_roles(cluster_permissions)
+
+          return [n for n in unknown_role_names if not n.endswith('-cluster')]
+
+     def _invalid_permission_names(self, subject_permission):
+          permissions = subject_permission.content['spec'].get('permissions', [])
+          role_names = [name for p in permissions if (name := p.get('clusterRoleName')) ]
+          unknown_role_names = self._filter_known_roles(role_names)
+
+          return [n for n in unknown_role_names if not n.endswith('-project')]
+
+     @staticmethod
+     def _filter_known_roles(role_names):
+          known_roles = ('admin', 'dedicated-readers', 'view')
+
+          return [n for n in role_names if not n in known_roles]
+
+
+_CLUSTER_ADMIN_NAMESPACE = 'openshift-backplane-cluster-admin'
+
+
+class _SubjectPermissionDenyClusterAdmin(_Rule):
+     def __init__(self, logger):
+          super().__init__('subject-permission-deny-cluster-admin', logger)
+
+     def run(self, registry):
+          failed = False
+
+          for path, subject_permission in registry.get_resources_of_type('SubjectPermission').items():
+               name = subject_permission.content['metadata']['name']
+
+               if invalid_permissions := self._invalid_permissions(subject_permission):
+                    failed = True
+
+                    self.log_failure(
+                         path,
+                         f"SubjectPermission '{name}' has permission(s) which grant"
+                         f" clusterRole(s) '{', '.join(invalid_permissions)}'"
+                         f" access to '{_CLUSTER_ADMIN_NAMESPACE}'"
+                    )
+
+          return failed
+
+     @staticmethod
+     def _invalid_permissions(subject_permission):
+          permissions = subject_permission.content['spec'].get('permissions', [])
+
+          invalid = []
+
+          for perm in permissions:
+               allowed_re = perm.get('namespacesAllowedRegex')
+               if not allowed_re or not re.match(allowed_re, _CLUSTER_ADMIN_NAMESPACE):
+                    continue
```

Comment:
This breaks the loop and will not check `namespacesDeniedRegex`

Reply:
Is that guideline that `namespacesDeniedRegex` must be defined and must exclude `'openshift-backplane-cluster-admin'`?

I interpreted it as if `namespacesAllowedRegex` is defined and matches `'openshift-backplane-cluster-admin' then the above must be true.

Reply:
Done.